### PR TITLE
Initial protocol discussion.

### PR DESCRIPTION
**Pull request by matthewtole**

### CODE REVIEW

File: message-header.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+This is the specification for the format and contents of a BAM header. 
+
+| Field Name | Type | Size (bytes) | Description |
+|------------|------|--------------|----------|
+| Message ID | int | 2 | The numerical ID of the message being sent. |
+| Message Size | int | 4 | How big the entire message is in bytes. |
+| Block Offset | int | 4 | The byte offset of this block of the message. |
+| Block Size | int | 2 | The size of this block in bytes. |
```
Comment (matthewtole) [inline message-header.md]: Do we need this? Doesn't AppMessage / Dictionary already tell us the size of the incoming packet?

File: message-header.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+This is the specification for the format and contents of a BAM header. 
+
+| Field Name | Type | Size (bytes) | Description |
+|------------|------|--------------|----------|
+| Message ID | int | 2 | The numerical ID of the message being sent. |
+| Message Size | int | 4 | How big the entire message is in bytes. |
+| Block Offset | int | 4 | The byte offset of this block of the message. |
```
Comment (matthewtole) [inline message-header.md]: If we have a fixed block size (for non-final blocks) then we could change this to be block position, which could potentially reduce the size of the field?

File: message-header.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+This is the specification for the format and contents of a BAM header. 
+
+| Field Name | Type | Size (bytes) | Description |
+|------------|------|--------------|----------|
+| Message ID | int | 2 | The numerical ID of the message being sent. |
+| Message Size | int | 3 | How big the entire message is in bytes. |
+| Block Offset | int | 4 | The byte offset of this block of the message. |
+| Block Size | int | 2 | The size of this block in bytes. |
```
Comment (matthewtole) [inline message-header.md]: Do we need this? Doesn't AppMessage / Dictionary already tell us the size of the incoming packet?

File: message-header.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+This is the specification for the format and contents of a BAM header. 
+
+| Field Name | Type | Size (bytes) | Description |
+|------------|------|--------------|----------|
+| Message ID | int | 2 | The numerical ID of the message being sent. |
+| Message Size | int | 3 | How big the entire message is in bytes. |
+| Block Offset | int | 4 | The byte offset of this block of the message. |
```
Comment (matthewtole) [inline message-header.md]: If we have a fixed block size (for non-final blocks) then we could change this to be block position, which could potentially reduce the size of the field?

**Comment by cynorg** [comment]
So, a couple things:
1) According to http://developer.getpebble.com/sdk/whats-new/#hardware-comparison - the app+heap memory space for Basalt is only 64kb (24kb for Aplite) - persistent storage can only hold 4k, and it's unlikely that an application doing anything useful (and including BAM) is going to fit under 4k, so I would imagine 64kb would be sufficient for the current version of the BAM protocol for both Aplite and Basalt.

2) That said, if we're designing a protocol, rather than just a message format, then we can track state/details on receiving ends, and don't need to send the full message size each time.  We should have data and control BAMs.  In fact, I'd suggest that the messaging handshake should be a single control message saying "I've got this many bytes, of this tentative type, set aside memory/object(s) to receive them" - and the watch can reserve that memory space then and there, and reply back, or reply back with a failure because there's insufficient free memory to allocate receiving.  Then we can easily enough reserve 3 or even 4 bytes ( ! so many bytes ! ) in the initial message without it "costing" much, to allow for a future when we can send that volume of data.

I don't think we want the control mechanism to be too heavy handed at all, essentially a create/destroy for individual messages, but it will allow for future expandability or options without making our data messages too thin.  Having a way to send small messages within BAM in a single packet would be useful, as well (perhaps having a 'data' field within a control packet, or having an implicit processing of single packets that are self-contained).

**Comment by matthewtole** [comment]
1. Good point. 2 bytes is definitely enough for the max size.
2. There will definitely be a tradeoff between sending an extra message and wasting a couple of bytes on each message.

**Pull request closed without merging**